docs/CONTROLFLOW.md

```diff
@@ -0,0 +1,343 @@
+func main() {
+	{
+		println("1")
+		goto _s2a_l3
+	}
+_s2a_l1:
+	{
+		println("3")
+		goto _s2a_l4
+	}
+_s2a_l2:
+	{
+		println("5")
+		return
+	}
+_s2a_l3:
+	{
+		println("2")
+		goto _s2a_l1
+	}
+_s2a_l4:
+	{
+		println("4")
+		goto _s2a_l2
+	}
+}
+ 
+```
+
+
+#### Junk jumps
+
+Param: `junk_jumps` (default: `0`)
+
+> Warning: this param affects resulting binary only when used in combination with [flattening](#control-flow-flattening)
+
+Junk jumps adds jumps to random blocks `junk_jumps` times, these inserted jumps can also form a chain. This param is useful for linearly increasing the functions complexity.
+
+Input:
+```go
+//garble:controlflow flatten_passes=0 junk_jumps=5 block_splits=0
+func main() {
+	if true {
+		println("hello world")
+	}
+}
+```
+
+Result: 
+
+```go
+func main() {
+	{
+		if true {
+			goto _s2a_l4
+		} else {
+			goto _s2a_l2
+		}
+	}
+_s2a_l1:
+	{
+		println("hello world")
+		goto _s2a_l7
+	}
+_s2a_l2:
+	{
+		return
+	}
+_s2a_l3:
+	{
+		goto _s2a_l2
+	}
+_s2a_l4:
+	{
+		goto _s2a_l6
+	}
+_s2a_l5:
+	{
+		goto _s2a_l3
+	}
+_s2a_l6:
+	{
+		goto _s2a_l1
+	}
+_s2a_l7:
+	{
+		goto _s2a_l5
+	}
+}
+```
+
+#### Control flow flattening
+
+Param: `flatten_passes` (default: `1`)
+
+
+This parameter completely [flattens the control flow](https://github.com/obfuscator-llvm/obfuscator/wiki/Control-Flow-Flattening) `flatten_passes` times, which makes analysing the logic of the function very difficult
+This is the most important parameter without which the other parameters have no effect on the resulting binary.
+
+> Warning: Unlike junk jumps, this parameter increases control flow complexity non-linearly. In most cases we do not recommend specifying a value greater than 3. Check [Benchmark](#complexity-benchmark)
+
+
+Input:
+```go
+//garble:controlflow flatten_passes=1 junk_jumps=0 block_splits=0
+func main() {
+	if true {
+		println("hello world")
+	} else {
+		println("not hello world")
+	}
+}
+```
+
+Result:
+
+```go
+func main() {
+	var _s2a_0 int
+_s2a_l0:
+	{
+		goto _s2a_l2
+	}
+_s2a_l1:
+	{
+		println("not hello world")
+		goto _s2a_l6
+	}
+_s2a_l2:
+	{
+		_s2a_1 := _s2a_0 == (int)(1)
+		if _s2a_1 {
+			goto _s2a_l9
+		} else {
+			goto _s2a_l8
+		}
+	}
+_s2a_l3:
+	{
+		_s2a_2 := _s2a_0 == (int)(4)
+		if _s2a_2 {
+			goto _s2a_l4
+		} else {
+			goto _s2a_l5
+		}
+	}
+_s2a_l4:
+	{
+		return
+	}
+_s2a_l5:
+	{
+		if true {
+			goto _s2a_l7
+		} else {
+			goto _s2a_l11
+		}
+	}
+_s2a_l6:
+	{
+		_s2a_0 = (int)(4)
+		goto _s2a_l0
+	}
+_s2a_l7:
+	{
+		_s2a_0 = (int)(1)
+		goto _s2a_l0
+	}
+_s2a_l8:
+	{
+		_s2a_3 := _s2a_0 == (int)(2)
+		if _s2a_3 {
+			goto _s2a_l1
+		} else {
+			goto _s2a_l12
+		}
+	}
+_s2a_l9:
+	{
+		println("hello world")
+		goto _s2a_l10
+	}
+_s2a_l10:
+	{
+		_s2a_0 = (int)(3)
+		goto _s2a_l0
+	}
+_s2a_l11:
+	{
+		_s2a_0 = (int)(2)
+		goto _s2a_l0
+	}
+_s2a_l12:
+	{
+		_s2a_4 := _s2a_0 == (int)(3)
+		if _s2a_4 {
+			goto _s2a_l4
+		} else {
+			goto _s2a_l3
+		}
+	}
+}
+```
+
+### Complexity benchmark
```

Comment:
nit: given how large the benchmark table below is, I would probably put this section at the very end of the doc, to not hide the caveats - which are actually pretty important.